pkg/client/group_mgmt_client.go

```diff
@@ -61,46 +61,63 @@ type Argument struct {
 	JobId string `json:"job_id,omitempty"`
 }
 
-// NewClient instantiates a new client to communicate with the Nimble group
-func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
-	// Create GroupMgmt Client
-	client := resty.New()
-	client.SetTLSClientConfig(&tls.Config{
+func newGroupMgmtClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) *GroupMgmtClient {
+	// Get new resty()
+	restyClient := resty.New()
+	restyClient.SetTLSClientConfig(&tls.Config{
 		InsecureSkipVerify: true,
 	})
+
+	// Create GroupMgmt Client
 	groupMgmtClient := &GroupMgmtClient{
 		URL:       fmt.Sprintf("https://%s:5392/%s", ipAddress, apiVersion),
-		Client:    client,
+		Client:    restyClient,
 		WaitOnJob: waitOnJobs,
 		Username:  username,
 		Password:  password,
 	}
+	return groupMgmtClient
+}
 
+// NewClient instantiates a new client to communicate with the Nimble group
+func NewClient(ipAddress, username, password, apiVersion string, waitOnJobs bool) (*GroupMgmtClient, error) {
+	// Get resty client
+	groupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
 	// Get session token
 	sessionToken, err := groupMgmtClient.login(username, password)
 	if err != nil {
 		return nil, err
 	}
 	groupMgmtClient.SessionToken = sessionToken
 
+	// Login retry counter
+	loginRetry := 0
 	// Add retryCondition
 	groupMgmtClient.Client.
-		SetRetryCount(maxOpsRetries).
 		AddRetryCondition(
 			func(resp *resty.Response, err error) bool {
+
 				if err == nil && resp.StatusCode() == 401 {
 					// login again and refreh the session token
-					sessionToken, err = groupMgmtClient.login(username, password)
-					if err != nil {
+					loginRetry++
+					if loginRetry == maxLoginRetries {
+						// exhausted the login attempt. backoff now .
 						return false
 					}
+					newGroupMgmtClient := newGroupMgmtClient(ipAddress, username, password, apiVersion, waitOnJobs)
```

Comment:
If I understand correctly, the reason to create a new client here is so we can perform a login without a retry condition.  Could we instead use the existing client and use a flag to indicate that we are performing a re-login, so the retry condition knows not to try another re-login?  It seems better to avoid creating extra client connections.

Reply:
Hey No, the groupMgmtClient.login(username, password)  function is getting called recursively on the same groupMgmtClient instance and it never returns if the password is invalid which is happening in this case. We need to have a new instance of groupMgmtClient to avoid the recursive call. Hope this clarifies.

Reply:
Your proposed logic will go in indefinite recursion and it will never reach this line "reloginInProgress = false". 
Even earlier implementation has similar issue that's the reason I have made the code change.

Reply:
If this is still not clear to you, I can set up a call and explain. thanks

Reply:
It is still not clear to me.  I thought the "!reloginInProgress" condition before the relogin attempt would prevent recursion.

FWIW, George's preference seems to be to avoid creating another client connection.  https://github.com/hpe-storage/nimble-golang-sdk/pull/12#discussion_r498813718

I can meet now if you want to set up a call.

Reply:
@rkumpf , made the suggested changes and seems to be working . Sorry i missed "reloginInProgress = true" in your porposed code . thanks